Foreman-bootloaders no longer ships binaries

lzap opened the pull request:
This is changing the way how we ship bootloaders in Red Hat. Instead of using pre-built binaries for different archs from upstream tarball, we generate them from grub module packages. This will only work in RHEL 7.4, they were recently added. This is a short term solution, longer term we will do this in puppet and we will most likely ditch this package which will kind of return to previous state. When I design this, there was no option to generate non-x86-86 grub bootloaders in x86_64 distro.

Installation and upgrade is clean and works, after installation files are created properly overwriting what was previously installed:

```
[root@hp-z600-01 ~]# ls /var/lib/tftpboot/grub2 -la
total 7768
drwxr-xr-x. 7 root root     247 Jan 24 08:54 .
drwxr-xr-x. 4 root root    4096 Jan 24 08:53 ..
drwxr-xr-x. 2 root root       6 Jan 24 08:53 fonts
-rw-r--r--. 1 root root  244835 Jan 24 08:54 grubia32.0
-rw-r--r--. 1 root root  685056 Jan 24 08:54 grubia32.efi
-rw-r--r--. 1 root root 1007616 Jan 24 08:54 grubx64.efi
drwxr-xr-x. 2 root root    8192 Jan 24 08:54 i386-efi
drwxr-xr-x. 2 root root    8192 Jan 24 08:54 i386-pc
drwxr-xr-x. 2 root root       6 Jan 24 08:53 locale
-rw-r--r--. 1 root root 1301312 Jan 24 08:54 shim.efi
-rw-r--r--. 1 root root 1035552 Jan 24 08:54 shimia32.efi
-rw-r--r--. 1 root root 1029400 Jan 24 08:54 shimia32-redhat.efi
-rw-r--r--. 1 root root 1301312 Jan 24 08:54 shimx64.efi
-rw-r--r--. 1 root root 1295152 Jan 24 08:54 shimx64-redhat.efi
drwxr-xr-x. 2 root root    8192 Jan 24 08:54 x86_64-efi
```

By default it only creates shim and grub2 for x86 and x86_64, because other platforms are only in RHEL (not in Fedora or CentOS). To generate other platforms, users must explicitly call the script:

    foreman-generate-bootloaders aa64
    foreman-generate-bootloaders ppc64

The first command fails currently, bug in RHEL: https://bugzilla.redhat.com/show_bug.cgi?id=1538084 but IBM POWER architectures (ppc, ppc64, ppc64le) all work fine.

* [x] Nightly

@ehelms @ares

ehelms commented on foreman-bootloaders-redhat/foreman-bootloaders-redhat.spec:
```diff
@@ -1,108 +1,105 @@
 
 %build
-
+cat > foreman-generate-bootloaders <<'EOF'
```
Would this be better as a script that is pulled in as a source? The script could still live with the packaging.

lzap replied:
Absolutely, rebased.

ares commented on foreman-bootloaders-redhat/foreman-bootloaders-redhat.spec:
```diff
@@ -1,108 +1,74 @@
 Requires: coreutils
 Requires: %{name}
 Requires: /var/lib/tftpboot
-Requires: foreman-proxy
```
if that is required for /var/lib/tftpboot, this should remain here

lzap replied:
It was required for some subdirectories to be present I think. I will keep it just to prevent any possible regression. Thanks.

ekohl commented on foreman-bootloaders-redhat/foreman-bootloaders-redhat.spec:
```diff
@@ -1,108 +1,74 @@
+URL: https://github.com/theforeman/foreman
 BuildArch: noarch
 
+Source0: foreman-generate-bootloaders
```
Forgot a git add?

lzap replied:
Indeed.

lzap commented:
Rebased, thanks for help here. This is quite high priority :)

ekohl commented on foreman-bootloaders-redhat/foreman-bootloaders-redhat.spec:
```diff
@@ -27,82 +30,46 @@ Requires: /var/lib/tftpboot
 
 
 %changelog
+* Wed Jan 24 2018 Lukas Zapletal <lzap+rpm@redhat.com> 201801241201-1
```
We don't keep changelogs - tito extracts them from git log and then generates them.

lzap replied:
Removing.

ekohl commented on foreman-bootloaders-redhat/foreman-generate-bootloaders:
```diff
@@ -0,0 +1,31 @@
+        [[ "$2" != "noinst" ]] && yum -y install grub2-efi-aa64-modules >/dev/null
+        generate arm64-efi grub2/grubaa64.efi
+elif [[ "$1" == "ppc" ]]; then
+        [[ "$2" != "noinst" ]] && yum -y remove grub2-ppc*-modules >/dev/null
```
What's the reasoning behind removing and then installing again?

lzap replied:
All ppc packages conflict between each other, so this solves the problem.

ekohl commented on foreman-bootloaders-redhat/foreman-generate-bootloaders:
```diff
@@ -0,0 +1,31 @@
+        grub2-mkimage -d /usr/lib/grub/$1/ -O $1 -o /var/lib/tftpboot/$2 -p "" $MODULES
+}
+if [[ "$1" == "aa64" ]]; then
+        [[ "$2" != "noinst" ]] && yum -y install grub2-efi-aa64-modules >/dev/null
```
Shouldn't this be a dependency?

lzap replied:
Unfortunately, aa64 and ppc* are not in CentOS, hard dependency would break it. Therefore all non-intel archs are explicit yum install. The option to skip installation is there because I call the script from yum transaction, unsure if yum in yum is allowed. I will redesign the options to be more explicit.

lzap commented:
Rebased, I decided to ditch all explicit yum package installs, now it warns you if you are missing a package. That is much more safer. Try now.

ekohl commented on foreman-bootloaders-redhat/foreman-generate-bootloaders:
```diff
@@ -0,0 +1,41 @@
+        rpm -q "$1" &>/dev/null || echo "Please install: yum -y install $1" >/dev/stderr
+}
+if [[ "$ARCH" == "aa64" ]]; then
+	check_pkg grub2-efi-aa64-modules
```
Mixing of tabs and spaces here

ekohl commented on foreman-bootloaders-redhat/foreman-generate-bootloaders:
```diff
@@ -0,0 +1,41 @@
+}
+
+check_pkg() {
+        rpm -q "$1" &>/dev/null || echo "Please install: yum -y install $1" >/dev/stderr
```
This doesn't error out - should it? Or should the script be ```#!/bin/bash -e```

lzap commented:
Created downstream BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1538255

lzap commented:
Retabbed, added `-e` as well.

ekohl commented:
I'm good with this PR if the mixing of tabs and spaces is fixed.

lzap commented:
Retabbed finally.

lzap commented:
Do not merge, minor problem on CentOS7, will update.

lzap commented:
Fixed for CentOS 7.4. It works:

```
[root@katello ~]# file /var/lib/tftpboot/grub2/* | sort
/var/lib/tftpboot/grub2/fonts:               directory
/var/lib/tftpboot/grub2/grubia32.0:          data
/var/lib/tftpboot/grub2/grubia32.efi:        PE32 executable (EFI application) Intel 80386 (stripped to external PDB), for MS Windows
/var/lib/tftpboot/grub2/grubx64.efi:         PE32+ executable (EFI application) x86-64 (stripped to external PDB), for MS Windows
/var/lib/tftpboot/grub2/i386-efi:            directory
/var/lib/tftpboot/grub2/i386-pc:             directory
/var/lib/tftpboot/grub2/locale:              directory
/var/lib/tftpboot/grub2/shim.efi:            PE32+ executable (EFI application) x86-64 (stripped to external PDB), for MS Windows
/var/lib/tftpboot/grub2/shimia32-centos.efi: PE32 executable (EFI application) Intel 80386 (stripped to external PDB), for MS Windows
/var/lib/tftpboot/grub2/shimia32.efi:        PE32 executable (EFI application) Intel 80386 (stripped to external PDB), for MS Windows
/var/lib/tftpboot/grub2/shimx64-centos.efi:  PE32+ executable (EFI application) x86-64 (stripped to external PDB), for MS Windows
/var/lib/tftpboot/grub2/shimx64.efi:         PE32+ executable (EFI application) x86-64 (stripped to external PDB), for MS Windows
/var/lib/tftpboot/grub2/x86_64-efi:          directory
```

On RHEL 7.4:

```
[root@winter ~]# file /var/lib/tftpboot/grub2/* | sort
/var/lib/tftpboot/grub2/fonts:               directory
/var/lib/tftpboot/grub2/grubaa64.efi:        PE32+ executable (EFI application) Unknown processor type 0xaa64 (stripped to external PDB), for MS Windows
/var/lib/tftpboot/grub2/grub.cfg:            ASCII text
/var/lib/tftpboot/grub2/grubia32.0:          data
/var/lib/tftpboot/grub2/grubia32.efi:        PE32 executable (EFI application) Intel 80386 (stripped to external PDB), for MS Windows
/var/lib/tftpboot/grub2/grubppc64.efi:       ELF 32-bit MSB executable, PowerPC or cisco 4500, version 1 (SYSV), statically linked, stripped
/var/lib/tftpboot/grub2/grubppc64.elf:       ELF 32-bit MSB executable, PowerPC or cisco 4500, version 1 (SYSV), statically linked, stripped
/var/lib/tftpboot/grub2/grubppc64le.efi:     ELF 32-bit MSB executable, PowerPC or cisco 4500, version 1 (SYSV), statically linked, stripped
/var/lib/tftpboot/grub2/grubppc64le.elf:     ELF 32-bit MSB executable, PowerPC or cisco 4500, version 1 (SYSV), statically linked, stripped
/var/lib/tftpboot/grub2/grubx64.efi:         PE32+ executable (EFI application) x86-64 (stripped to external PDB), for MS Windows
/var/lib/tftpboot/grub2/i386-efi:            directory
/var/lib/tftpboot/grub2/i386-pc:             directory
/var/lib/tftpboot/grub2/locale:              directory
/var/lib/tftpboot/grub2/shimaa64.efi:        MS-DOS executable
/var/lib/tftpboot/grub2/shimaa64-redhat.efi: MS-DOS executable
/var/lib/tftpboot/grub2/shim.efi:            PE32+ executable (EFI application) x86-64 (stripped to external PDB), for MS Windows
/var/lib/tftpboot/grub2/shimx64.efi:         PE32+ executable (EFI application) x86-64 (stripped to external PDB), for MS Windows
/var/lib/tftpboot/grub2/shimx64-redhat.efi:  PE32+ executable (EFI application) x86-64 (stripped to external PDB), for MS Windows
/var/lib/tftpboot/grub2/x86_64-efi:          directory
```

It's now fixed!

lzap commented on foreman-bootloaders-redhat/foreman-generate-bootloaders:
```diff
@@ -0,0 +1,52 @@
+
+check_pkg shim-ia32
+check_pkg shim-x64
+cp -f /boot/efi/EFI/*/shim*.efi /var/lib/tftpboot/grub2
```
This was the only change, directories are different on OSes: `/boot/efi/EFI/redhat` vs `/boot/efi/EFI/centos`

lzap commented on foreman-bootloaders-redhat/foreman-generate-bootloaders:
```diff
@@ -0,0 +1,52 @@
+}
+
+check_pkg() {
+  rpm -q "$1" &>/dev/null || echo "Please install: yum -y install $1" >/dev/stderr && exit 1
```
And one more change, call `exit` when package is missing.

lzap commented:
Many thanks to @ehelms and @ekohl for fast review.
